rust/kernel/xarray.rs

```diff
@@ -0,0 +1,296 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! XArray abstraction.
+//!
+//! C header: [`include/linux/xarray.h`](../../include/linux/xarray.h)
+
+use crate::{
+    bindings,
+    error::{Error, Result},
+    types::{ForeignOwnable, Opaque},
+};
+use core::{marker::PhantomData, ops::Deref};
+
+/// Flags passed to `XArray::new` to configure the `XArray`.
+type Flags = bindings::gfp_t;
+
+/// Flag values passed to `XArray::new` to configure the `XArray`.
+pub mod flags {
+    /// Use IRQ-safe locking
+    pub const LOCK_IRQ: super::Flags = bindings::BINDINGS_XA_FLAGS_LOCK_IRQ;
+    /// Use softirq-safe locking
+    pub const LOCK_BH: super::Flags = bindings::BINDINGS_XA_FLAGS_LOCK_BH;
+    /// Track which entries are free (distinct from None)
+    pub const TRACK_FREE: super::Flags = bindings::BINDINGS_XA_FLAGS_TRACK_FREE;
+    /// Initialize array index 0 as busy
+    pub const ZERO_BUSY: super::Flags = bindings::BINDINGS_XA_FLAGS_ZERO_BUSY;
+    /// Use GFP_ACCOUNT for internal memory allocations
+    pub const ACCOUNT: super::Flags = bindings::BINDINGS_XA_FLAGS_ACCOUNT;
+    /// Create an allocating `XArray` starting at index 0
+    pub const ALLOC: super::Flags = bindings::BINDINGS_XA_FLAGS_ALLOC;
+    /// Create an allocating `XArray` starting at index 1
+    pub const ALLOC1: super::Flags = bindings::BINDINGS_XA_FLAGS_ALLOC1;
+}
+
+/// Wrapper for a value owned by the `XArray` which holds the `XArray` lock until dropped.
+///
+/// # Invariants
+///
+/// The `*mut T` is always non-NULL and owned by the referenced `XArray`
+pub struct Guard<'a, T: ForeignOwnable>(*mut T, &'a Opaque<bindings::xarray>);
+
+impl<'a, T: ForeignOwnable> Guard<'a, T> {
+    /// Borrow the underlying value wrapped by the `Guard`.
+    ///
+    /// Returns a `T::Borrowed` type for the owned `ForeignOwnable` type.
+    pub fn borrow<'b>(&'b self) -> T::Borrowed<'b>
+    where
+        'a: 'b,
+    {
+        // SAFETY: The value is owned by the `XArray`, the lifetime it is borrowed for must not
+        // outlive the `XArray` itself, nor the Guard that holds the lock ensuring the value
+        // remains in the `XArray`.
+        unsafe { T::borrow(self.0 as _) }
+    }
+}
+
+// Convenience impl for `ForeignOwnable` types whose `Borrowed`
+// form implements Deref.
+impl<'a, T: ForeignOwnable> Deref for Guard<'a, T>
+where
+    T::Borrowed<'static>: Deref,
+{
+    type Target = <T::Borrowed<'static> as Deref>::Target;
+
+    fn deref(&self) -> &Self::Target {
+        // SAFETY: See the `borrow()` method. The dereferenced `T::Borrowed` value
+        // must share the same lifetime, so we can return a reference to it.
+        // TODO: Is this really sound?
```

Comment:
Yeah, this is very interesting. I tried a few lifetime tricks, and couldn't make it work. @bjorn3 @nbdd0121 @wedsonaf anything you guys can think of? I think the intention here is to express that the lifetime of `Borrowed` outlives the lifetime of `&Self::Target`. But I may be wrong...

Reply:
This could be written as:
```rust
impl<'a, T: ForeignOwnable> Deref for Guard<'a, T>
where
    T::Borrowed<'static>: Deref,
    for<'b> T::Borrowed<'b>: Into<&'b <T::Borrowed<'static> as Deref>::Target>,
{
    type Target = <T::Borrowed<'static> as Deref>::Target;

    fn deref(&self) -> &Self::Target {
        self.borrow().into()
    }
}
```
then we just need to ensure `ArcBorrow<'a, T>: Into<'a T>` for this to be usable for `Arc`.

Reply:
Current code, as is, isn't really sound because `T::Borrowed<'self>: Deref` can do something funky inside and return a pointer to itself, rather than the thing it points to. What we really want here is a way to convert `T::Borrowed<'self>` to `&'self Target`, so it seems to me that just adding an explicit `Into` is the best way.

Reply:
Yeah, that was my feeling when writing this too (that it's safe given *existing* `T::Borrowed` implementations that follow certain rules but not sound in the general case). I'll add an `Into` impl to `ArcBorrow` for the next revision and see how that goes, thanks! @nbdd0121 do you want to be added as `Co-developed-by`?

Reply:
I couldn't figure this out... `Into` conflicts with the blanket `From` impl.

Since this is just a convenience, I think we can leave it for a future patch, so I've dropped this for now ^^;;